taichi/ir/frontend_ir.cpp

```diff
@@ -721,28 +741,70 @@ bool IndexExpression::is_global() const {
   return is_field() || is_matrix_field() || is_ndarray();
 }
 
+static void field_validation(FieldExpression *field_expr, int index_dim) {
+  TI_ASSERT(field_expr != nullptr);
+  TI_ASSERT(field_expr->snode != nullptr);
+  int field_dim = field_expr->snode->num_active_indices;
+
+  if (field_dim != index_dim) {
+    throw TaichiIndexError(
+        fmt::format("Field with dim {} accessed with indices of dim {}",
+                    field_dim, index_dim));
+  }
+}
+
+static void matrix_field_validation(MatrixFieldExpression *matrix_field_expr,
+                                    int index_dim) {
+  TI_ASSERT(matrix_field_expr != nullptr);
+  TI_ASSERT(!matrix_field_expr->fields.empty());
+  auto field_expr = matrix_field_expr->fields[0].cast<FieldExpression>();
+  int element_dim = matrix_field_expr->element_shape.size();
```

Comment:
For matrix field there's no need to consider `element_dim`.

Reply:
I was thinking we have to handle the following two cases, one of which has element_dim involved?

```
x = ti.Vector.field(n=2, shape=(10, 20, 30))

@ti.kernel
def test():
  x[0, 1, 2][0] = 1   # outter_dim + element_dim == index_dim
  x[0, 1, 2] = [10, 20] # outter_dim == index_dim
```

Reply:
Here `x[0, 1, 2][0]` is always becoming `IndexExpr(IndexExpr(x, [0, 1, 2]), [0])`.

Reply:
Ah that's right~